### PR TITLE
Feat: add register

### PR DESCRIPTION
**Pull request by Thuan2k3**

### CODE REVIEW

File: demo/src/main/java/com/jwtproject/demo/Entity/UserInfo.java [hunk 1]
```diff
@@ -18,6 +18,7 @@ public class UserInfo {
     private String name;
     private String email;
```
Comment (bwv-kiet-vh) [inline demo/src/main/java/com/jwtproject/demo/Entity/UserInfo.java:19]: What if 2 users have the same email?

File: demo/src/main/java/com/jwtproject/demo/Entity/UserInfo.java [hunk 2]
```diff
@@ -18,6 +18,7 @@ public class UserInfo {
...
     private String password;
+    private String phoneNumber;
     private String roles;
```
Comment (bwv-kiet-vh) [inline demo/src/main/java/com/jwtproject/demo/Entity/UserInfo.java:22]: Need to add attributes such as data type, length to the database, do not use default

Comment (bwv-kiet-vh) [reply demo/src/main/java/com/jwtproject/demo/Entity/UserInfo.java:22]: Example   @Column(name = "roles", columnDefinition = "integer default 1")

File: demo/src/main/java/com/jwtproject/demo/Service/UserInfoService.java [hunk 3]
```diff
@@ -32,10 +32,10 @@ public UserDetails loadUserByUsername(String username) throws UsernameNotFoundEx
                 .orElseThrow(() -> new UsernameNotFoundException("User not found " + username));
     }
 
-    public String addUser(UserInfo userInfo) {
+    public boolean addUser(UserInfo userInfo) {
         userInfo.setPassword(encoder.encode(userInfo.getPassword()));
         repository.save(userInfo);
-        return "User Added Successfully";
+        return true;
     }
```
Comment (bwv-kiet-vh) [inline demo/src/main/java/com/jwtproject/demo/Service/UserInfoService.java:39]: How do you know it always returns true?

Are there any cases where saving fails?

File: demo/src/main/java/com/jwtproject/demo/Entity/RegisterRequest.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    @Size(max = 50, message = "Name must be at most 50 characters long")
+    private String name;
+
+    @NotBlank(message = "Password is required")
+    @Size(max = 20, message = "Password must be at most 20 characters long")
+    @Size(min = 8, message = "Password must be at least 8 characters long")
+    private String password;
+
+    @NotBlank(message = "Email is required")
+    @Email(message = "Email is not valid", regexp = "^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$")
```
Comment (bwv-kiet-vh) [inline demo/src/main/java/com/jwtproject/demo/Entity/RegisterRequest.java:20]: Use MailRFC

File: demo/src/main/java/com/jwtproject/demo/Controller/UserController.java [hunk 5]
```diff
@@ -44,9 +48,43 @@ public String welcome() {
...
+                responseObject.setMessage(Message.FAIL);
+                responseObject.setStatusCode(HttpStatus.BAD_REQUEST.value());
+            }
+        } catch (DataIntegrityViolationException exception) {
+            responseObject = ReponseObject.builder()
+                    .message(Message.VALIDATION_ERRORS)
+                    .statusCode(HttpStatus.BAD_REQUEST.value())
+                    .data(Map.of("email", Message.EMAIL_EXIST))
+                    .build();
+        }
```
Comment (bwv-kiet-vh) [inline demo/src/main/java/com/jwtproject/demo/Controller/UserController.java:86]: This catch is only if there are duplicate emails in the database,
I didn't see you check this?

File: demo/src/main/java/com/jwtproject/demo/Controller/UserController.java [hunk 6]
```diff
@@ -44,9 +48,43 @@ public String welcome() {
         return "Welcome this endpoint is not secure";
     }
 
+//    @PostMapping("/addNewUser")
+//    public String addNewUser(@RequestBody UserInfo userInfo) {
+//        return service.addUser(userInfo);
+//    }
```
Comment (bwv-kiet-vh) [inline demo/src/main/java/com/jwtproject/demo/Controller/UserController.java:54]: Remove

File: demo/src/main/java/com/jwtproject/demo/Controller/UserController.java [hunk 7]
```diff
@@ -44,9 +48,43 @@ public String welcome() {
...
     @PostMapping("/addNewUser")
```
Comment (bwv-kiet-vh) [inline demo/src/main/java/com/jwtproject/demo/Controller/UserController.java:55]: @PostMapping("/auth/register")

File: demo/src/main/java/com/jwtproject/demo/Controller/UserController.java [hunk 8]
```diff
@@ -21,6 +24,7 @@
 import org.springframework.security.core.userdetails.UsernameNotFoundException;
 import org.springframework.web.bind.annotation.*;
 
+import java.util.Map;
 import java.util.Optional;
```
Comment (bwv-kiet-vh) [inline demo/src/main/java/com/jwtproject/demo/Controller/UserController.java:29]: @RequestMapping("/api/v1")

File: demo/src/main/java/com/jwtproject/demo/Controller/UserController.java [hunk 9]
```diff
@@ -44,9 +48,43 @@ public String welcome() {
...
+            newUser.setRoles(registerRequest.getRoles());
+
+            // Handle user creation and update ResponseObject
+            boolean isCreated = service.addUser(newUser);
+            if (isCreated) {
+                responseObject.setMessage(Message.SUCCESS);
+                return new ResponseEntity<>(responseObject, HttpStatus.OK);
+            } else {
+                responseObject.setMessage(Message.FAIL);
+                responseObject.setStatusCode(HttpStatus.BAD_REQUEST.value());
```
Comment (bwv-kiet-vh) [inline demo/src/main/java/com/jwtproject/demo/Controller/UserController.java:78]: I see created always returns true, so when does this case occur?

File: demo/src/main/java/com/jwtproject/demo/Entity/RegisterRequest.java (hunk 4 above)
Comment (Thuan2k3) [reply demo/src/main/java/com/jwtproject/demo/Entity/RegisterRequest.java:20]: I think it is also effective as MailRFC

File: demo/src/main/java/com/jwtproject/demo/Entity/AuthRequest.java [hunk 10]
```diff
@@ -9,9 +12,10 @@
 @NoArgsConstructor
 public class AuthRequest {
 
-
-    @Email
+    @NotBlank(message = "Email is required")
+    @MailRFCConstraint
     private String username;
```
[the remark this answers is not in the document]
Comment (Thuan2k3) [reply demo/src/main/java/com/jwtproject/demo/Entity/AuthRequest.java:17]: OK, I will fix it

File: demo/src/main/java/com/jwtproject/demo/Entity/UserInfo.java [hunk 11]
```diff
@@ -15,9 +14,23 @@ public class UserInfo {
...
+    @Column(name = "password")
+    @Length(max = 255)
     private String password;
-    private String roles;
 
+    @Column(name = "phoneNumber")
+    @Length(max = 15)
+    private String phoneNumber;
+
+    @Column(name = "roles", columnDefinition = "varchar(50) default \"ROLE_USER\"")
```
[the remark this answers is not in the document]
Comment (Thuan2k3) [reply demo/src/main/java/com/jwtproject/demo/Entity/UserInfo.java:34]: It will accept that

**Comment by thekids1002** [comment]
merge đại

**Pull request merged**